jsonschema/validators.py

```diff
@@ -862,6 +868,44 @@ def resolve_remote(self, uri):
             self.store[uri] = result
         return result
 
+    def store_subschema(self, schema, last_schema=None, last_url=None):
+        """
+        Using $id or id with $ref, save subschema to self.store
+
+        Arguments:
+
+            schema:
+
+                The referring schema.
+
+            last_schema:
+
+                The referring upper level schema.
+
+            last_url:
+
+                Save the last URL.
+        """
+        if not isinstance(schema, dict) \
+                or self.resolution_scope in list_schema:
+            return
+
+        for k in schema.keys():
+            if k in [u"id", u"$id"] and isinstance(schema[k], str_types):
+
+                last_url = urljoin(last_url, schema[k], allow_fragments=True)
```

Comment:
Splice the value and the url of the upper level, and save it into last_url at the same time.